### PR TITLE
New optional parameter (selectPaths) for query specs

### PR DESCRIPTION
**Pull request by wmathurin**
If you specify [path1, path2] you get back [soupElt[path1], soupElt[path2]] for each matching row instead of the soupElt's.

This pull request has the native code changes and tests.
Hybrid and react native support will come in a separate pull request.

### CODE REVIEW

File: libs/SmartStore/SmartStore/Classes/SFQuerySpec.h [hunk 1]
```diff
@@ -142,32 +148,37 @@ typedef NS_ENUM(NSUInteger, SFSoupQuerySortOrder) {
...
  * @param soupName The target soup name.
+ * @param selectPaths The paths to return - if nil the entire soup element is returned.
  * @param path The path to filter on.
  * @param matchKey The exact value to match.
  * @param orderPath The path to sort by.
  * @param order The sort order.
  * @param pageSize The page size.
  * @return A query spec object.
  */
++ (SFQuerySpec*) newExactQuerySpec:(NSString*)soupName withSelectPaths:(NSArray*)selectPaths withPath:(NSString*)path withMatchKey:(NSString*)matchKey withOrderPath:(NSString*)orderPath withOrder:(SFSoupQuerySortOrder)order withPageSize:(NSUInteger)pageSize;
```
Comment (wmathurin) [inline libs/SmartStore/SmartStore/Classes/SFQuerySpec.h]: Keeping old apis

File: libs/SalesforceSDKCore/SalesforceSDKCoreTests/SFSDKCryptoUtilsTests.m [hunk 2]
```diff
@@ -44,7 +44,7 @@ - (void)testRandomDataGenerator
     
     for (NSUInteger i = 0; i < numDataStrings; i++) {
         for (NSUInteger j = i + 1; j < numDataStrings; j++) {
-            XCTAssertFalse([[dataStringArray objectAtIndex:i] isEqualToData:[dataStringArray objectAtIndex:j]], @"Random data strings at index %lu and %d are equal.  Not enough entropy!", (unsigned long)i, j);
+            XCTAssertFalse([[dataStringArray objectAtIndex:i] isEqualToData:[dataStringArray objectAtIndex:j]], @"Random data strings at index %lu and %luu are equal.  Not enough entropy!", (unsigned long)i, (unsigned long)j);
```
Comment (wmathurin) [inline libs/SalesforceSDKCore/SalesforceSDKCoreTests/SFSDKCryptoUtilsTests.m]: Taking care of some of the warnings

File: libs/SmartStore/SmartStore/Classes/SFQuerySpec.m [hunk 3]
```diff
@@ -246,16 +271,19 @@ - (NSString*) computeIdsSql:(NSString*) smartSql {
 }
 
 - (NSString*)computeSelectClause {
-    return [@[@"SELECT ", [self computeFieldReference:@"_soup"], @" "] componentsJoinedByString:@""];
+    NSMutableArray* fieldReferences = [NSMutableArray new];
+    for (NSString* selectPath in (self.selectPaths ? self.selectPaths : @[@"_soup"])) {
```
Comment (wmathurin) [inline libs/SmartStore/SmartStore/Classes/SFQuerySpec.m]: That's where it's all happening

File: libs/SmartStore/SmartStore/Classes/SFQuerySpec.m [hunk 4]
```diff
@@ -296,10 +318,19 @@ - (NSString*)computeWhereClause {
 
         case kSFSoupQueryTypeMatch:
             return [@[@"WHERE ",
-                      [self computeSoupFtsReference], @".", DOCID_COL, @" = ", [self computeFieldReference:SOUP_ENTRY_ID], // join clause
-                      @" AND ",
-                      field, @" MATCH '", self.matchKey, @"' "  // match clause -- statement arg binding doesn't seem to work so inlining matchKey
-                      ] componentsJoinedByString:@""];
+                      [self computeFieldReference:SOUP_ENTRY_ID],
```
Comment (wmathurin) [inline libs/SmartStore/SmartStore/Classes/SFQuerySpec.m]: See description of https://github.com/forcedotcom/SalesforceMobileSDK-iOS/pull/1568/commits/95486afffcdc236de4161aa473dad70bc9df2180 for more information

**Comment by bhariharan** [comment]
👍

**Pull request merged**